### ISSUE TITLE
Select the largest frame of ICO files

### ISSUE THREAD

**Issue by db-one** [d2phap/ImageGlass#994]
![image](https://user-images.githubusercontent.com/20243226/109802471-d2c5a080-7c5a-11eb-8cf6-695cf4bc5008.png)

Hi, can you add support for ICO icon files, for example, this icon contains seven icons of different sizes. If you use the WINDOWS default file manager, you can browse the seven icons separately, but ImageGlass can only display the first one by default , And some icons are displayed abnormally, such as the following, so can you add a button to switch the icons in the ICO file and browse normally, thank you very much

![snipaste20210303_195438](https://user-images.githubusercontent.com/20243226/109802287-9134f580-7c5a-11eb-849c-1df0aeb33a82.jpg)

**Comment by d2phap** [comment]
You can use Page Navigation tool:
![image](https://user-images.githubusercontent.com/3154213/109816694-fba26180-7c6b-11eb-9be0-58f813e44199.png)

**Comment by db-one** [comment]
![image](https://user-images.githubusercontent.com/20243226/109907624-a1da7f80-7cdd-11eb-8914-bd3fc578d1c5.png)
Okay, thank you very much for your reply, can you add these two options to the custom toolbar

**Comment by db-one** [comment]
Okay, thank you very much

**Comment by db-one** [comment]
![image](https://user-images.githubusercontent.com/20243226/109910227-9c336880-7ce2-11eb-8881-2a3e716dc25d.png)

![image](https://user-images.githubusercontent.com/20243226/109910029-31822d00-7ce2-11eb-8930-d23f3c221445.png)

For example, the size of the image opened by ImageGlass will be displayed incompletely as shown below. Below is the original file of the icon for reference, thank you

![image](https://user-images.githubusercontent.com/20243226/109909966-08fa3300-7ce2-11eb-8e8d-f48daf1a3d8f.png)



[SonyMDRV700.zip](https://github.com/d2phap/ImageGlass/files/6080798/SonyMDRV700.zip)
[photos.zip](https://github.com/d2phap/ImageGlass/files/6080805/photos.zip)

**Comment by K0media** [comment]
@d2phap, acutally, I'm having an issue about how the icons are handled.
Please see below:

https://user-images.githubusercontent.com/22375465/114242091-b22aed80-9960-11eb-8b50-3304e7e992e7.mp4

What can be done about this? Any workarounds or it works that way?
Is there another issue open regarding this problem? I thought of opening an issue, but I think it would be a duplicate.

Is it going to be fixed by default any time soon?

Thanks.

**Comment by fire-eggs** [comment]
@K0media :
>acutally, I'm having an issue about how the icons are handled. Please see below:

Could you state what your actual issue is? I see you clicking on what are presumably icon files, zooming in and out, and clicking on another icon.

What is it about the ImageGlass behavior which is unacceptable?

**Comment by K0media** [comment]
> @K0media :
> 
> > acutally, I'm having an issue about how the icons are handled. Please see below:
> 
> Could you state what your actual issue is? I see you clicking on what are presumably icon files, zooming in and out, and clicking on another icon.
> 
> What is it about the ImageGlass behavior which is unacceptable?

It's simple. It is handled abnormally by IG, while it displays fine on other media viewers.

I'll record another screencast to show you. (I'd record a video with audio explaining details, but that would take longer. I can do it if necessary later, though.)

https://user-images.githubusercontent.com/22375465/114246897-39309380-996a-11eb-8092-8d47c005dca6.mp4

As you can see above, some icons display fine in "normal" resolution, but some others are distorted and pixelated that we almost can't see the proper format.

And, when opening with another app, such as Windows 10 default Photos app, it displays fine with the anti-aliasing effect.

There should be something wrong here... I tried customizing options, but it's still mishandled by IG.

Also, I can share the icon pack if needed, so you can research better on it.

**Comment by d2phap** [comment]
@K0media 
You can change Zoom Optimization option to Smooth pixels (in v8.0), or Low quality (in v8.1) in Settings > Image > Zooming

**Comment by fire-eggs** [comment]
Right, I understand now. Thanks for the detailed videos!

TLDR; sorry to say but this is the current state of ImageGlass, no workarounds possible.

Some of these other apps presumably are taking advantage of the different sized icons contained in the ICO. [E.g. as you zoom in, the apps use the next larger icon file as necessary.] ImageGlass (currently) does not: the first, smallest icon file is the only one used and zooming just gets big pixels.

I'm not quite sure what's going on with the anti-aliasing. I.e. is that a feature of the other media apps or something in the newest icon formats that I'm unaware of? Either way, ImageGlass isn't currently supporting it.

OK, from wikipedia:
>The Windows Vista Explorer supports smoothly scaling icons to non-standard sizes which are >rendered on the fly even if an image is not present for that size in the icon file.

ImageGlass is not doing this.

**Comment by iDm1** [comment]
I think you're overcomplicating how other viewers are handling ICO files. This format usually contains several different images for different sizes. Most image viewers are displaying the largest of them by default, but ImageGlass always displays the first one, which is often the smallest one. 

Displaying the largest image from ICO file by default is enough. Dynamic switching to the closest size is nice of course, but it requires much more changes in ImageGlass to implement it.

**Comment by db-one** [comment]
> I think you're overcomplicating how other viewers are handling ICO files. This format usually contains several different images for different sizes. Most image viewers are displaying the largest of them by default, but ImageGlass always displays the first one, which is often the smallest one.
> 
> Displaying the largest image from ICO file by default is enough. Dynamic switching to the closest size is nice of course, but it requires much more changes in ImageGlass to implement it.

Yes, Strongly Agree

**Comment by fire-eggs** [comment]
>Most image viewers are displaying the largest of them by default, but ImageGlass always displays the first one, which is often the smallest one.

My thanks for the feedback! My ignorance of the ICO format is showing ...

**Comment by d2phap** [comment]
I spent my time today to refactor the code for ICO file, so now IG will display the largest frame of the viewing ICO file by default.
However, the problem with displaying (picture below) is not fixed. ImageGlass is using Magick.NET for loading ICO, apparently Magick.NET has issue with this ico file.

This change will be included in the next official release (soon).

![image](https://user-images.githubusercontent.com/3154213/118353532-4b20da00-b599-11eb-9bc0-84b8c1c20aca.png)

**Comment by d2phap** [comment]
released v8.2!
